File: platform/android/studio/app/src/main/java/org/fullstacked/editor/Adapter.kt
package org.fullstacked.editor

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.io.File
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.attribute.BasicFileAttributes


open class Adapter(
    val projectId: String,
    baseDirectory: String
) {
    val platform = "android"
    val fs: AdapterFS = AdapterFS(baseDirectory)

    open fun getFile(path: String) : InputStream? {
        val data = this.fs.readFile(path, false);
        if(data is AdapterError) return null
        return ByteArrayInputStream(data as ByteArray)
    }

    fun writeFile(path: String, data: ByteArray?, recursive: Boolean) : Any? {
        if(recursive) {
            val dir = path.split("/").dropLast(1)
            this.fs.mkdir(dir.joinToString("/"))
        }

        return this.fs.writeFile(path, data ?: byteArrayOf())
    }

    open fun callAdapterMethod(methodPath: ArrayList<String>, body: String?): Any? {
        if(methodPath.isEmpty()) return null

        val json = if(!body.isNullOrEmpty()) JSONArray(body) else JSONArray("[]")

        when (methodPath.first()) {
            "platform" -> return this.platform
            "fs" -> return this.fsSwitch(methodPath[1], json)
            "fetch" -> return this.fetch(json)
            "broadcast" -> return this.broadcast(json)
        }

        return null
    }

    private fun broadcast(json: JSONArray) : Boolean {
        val peerMessage = JSONObject()
        peerMessage.put("projectId", this.projectId)
        peerMessage.put("data", json.getString(0))
        InstanceEditor.singleton.push("sendData", peerMessage.toString())

        return true
    }

    private fun fsSwitch(method: String, json: JSONArray) : Any? {
        when (method) {
            "readFile" -> {
                var utf8 = false
                if(json.length() > 1) {
                    val opt = json.getJSONObject(1)
                    try {
                        utf8 = opt.getString("encoding") == "utf8"
                    } catch (_: Exception) { }
                }
                return this.fs.readFile(json.getString(0), utf8)
            }
            "writeFile" -> {
                var data: ByteArray? = null

                try {
                    if(json.getJSONObject(1).get("type") == "Uint8Array") {
                        val numberArr = json.getJSONObject(1).getJSONArray("data")
                        val byteArray = ByteArray(numberArr.length())
                        for (i in 0..<numberArr.length()) {
                            byteArray[i] = numberArr.get(i).toString().toInt().toByte()
                        }
                        data = byteArray
                    }
                }catch (e: Exception) {
                    data = json.getString(1).toByteArray()
                }

                var recursive = false
                if(json.length() > 2) {
                    val opt = json.getJSONObject(2)
                    try {
                        recursive = opt.getBoolean("recursive")
                    } catch (_: Exception) { }
                }

                return this.writeFile(json.getString(0), data, recursive)
            }
            "writeFileMulti" -> {
                var recursive = false
                if(json.length() > 1) {
                    val opt = json.getJSONObject(1)
                    try {
                        recursive = opt.getBoolean("recursive")
                    } catch (_: Exception) { }
                }

                val files = json.getJSONArray(0)
                for (i in 0..<files.length()) {
                    val file = files.getJSONObject(i)
                    var data: ByteArray? = null

                    try {
                        val uint8array = file.getJSONObject("data")
                        if(uint8array.get("type") == "Uint8Array") {
                            val numberArr = uint8array.getJSONArray("data")
                            val byteArray = ByteArray(numberArr.length())
                            for (index in 0..<numberArr.length()) {
                                byteArray[index] = numberArr.get(index).toString().toInt().toByte()
                            }
                            data = byteArray
                        }
                    }catch (e: Exception) {
                        data = file.getString("data").toByteArray()
                    }

                    val maybeError = this.writeFile(file.getString("path"), data, recursive)
                    if(maybeError != null && maybeError != true)
                        return maybeError
                }

                return true
            }
            "unlink" -> return this.fs.unlink(json.getString(0))
            "readdir" -> {
                var recursive = false
                var withFileTypes = false
                if(json.length() > 1) {
                    val opt = json.getJSONObject(1)
                    try {
                        recursive = opt.getBoolean("recursive")
                    } catch (_: Exception) { }
                    try {
                        withFileTypes = opt.getBoolean("withFileTypes")
                    } catch (_: Exception) { }
                }

                return this.fs.readdir(json.getString(0), withFileTypes, recursive)
            }
            "mkdir" -> return this.fs.mkdir(json.getString(0))
            "rmdir" -> return this.fs.rmdir(json.getString(0))
            "stat" -> return this.fs.stat(json.getString(0))
            "lstat" -> return this.fs.stat(json.getString(0))
            "exists" -> return this.fs.exists(json.getString(0))
        }

        return null
    }

    private fun fetch(json: JSONArray): Any? {
        val client = OkHttpClient()
        val request = Request.Builder()
            .url(json.getString(0))

        var utf8 = false
        var method = "GET"
        val headers = mutableMapOf<String, String>()
        var body: ByteArray? = null

        if(json.length() > 1) {
            val opt = json.getJSONObject(1)

            try {
                utf8 = opt.getString("encoding") == "utf8"
            } catch (_: Exception) { }

            try {
                method = opt.getString("method")
            } catch (_: Exception) { }

            try {
                val jsonHeaders = opt.getJSONObject("headers")
                for(key in jsonHeaders.keys()) {
                    headers[key] = jsonHeaders.getString(key)
                }
            } catch (_: Exception) { }

            try {
                val jsonBody = opt.getJSONObject("body")
                if(jsonBody.getString("type") == "Uint8Array") {
                    val numberArr = jsonBody.getJSONArray("data")
                    val byteArray = ByteArray(numberArr.length())
                    for (i in 0..<numberArr.length()) {
                        val numStr = numberArr.get(i).toString()
                        val num = numStr.toInt()
                        byteArray[i] = num.toByte()
                    }
                    body = byteArray
                }
            } catch (_: Exception) { }

            if(body == null) {
                try {
                    val jsonBody = opt.getString("body")
                    body = jsonBody.toByteArray()
                } catch (_: Exception) { }
            }
        }

        request.method(method, body?.toRequestBody())

        headers.forEach { (key, value) ->
            request.addHeader(key, value)
        }

        if(body != null) {
            request.addHeader("content-length", body.size.toString())
        }

        val response: Response = try {
            client.newCall(request.build()).execute()
        } catch (e: Exception) { return null }

        val responseJson = JSONObject()

        responseJson.put("statusCode", response.code)
        responseJson.put("statusMessage", response.message)

        val responseHeaders = JSONObject()
        response.headers.forEach { (key, value) ->
            responseHeaders.put(key, value)
        }
        responseJson.put("headers", responseHeaders)

        if(utf8) {
            if(response.body == null) {
                responseJson.put("body", "")
            } else {
                responseJson.put("body", response.body?.string() ?: "")
            }
        } else {
            val responseBody = JSONObject()
            responseBody.put("type", "Uint8Array")

            val responseBodyData = JSONArray()
            if(response.body != null) {
                response.body?.bytes()?.forEach { byte ->
                    responseBodyData.put(byte)
                }
            }
            responseBody.put("data", responseBodyData)

            responseJson.put("body", responseBody)
        }

        return responseJson
    }
}

data class AdapterError(
    val code: String,
    val path: String,
    val syscall: String,
)

class AdapterFS(private val baseDirectory: String) {
    // @return null if doesn't exist, true if exists and directory, false if exists and is file
    private fun itemExistsAndIsDirectory (path: String) : Boolean? {
        val item = File(path)
        if(!item.exists()) return null
        return item.isDirectory
    }

    fun readFile(path: String, utf8: Boolean) : Any {
        val itemPath = this.baseDirectory + "/" + path

        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == null || existsAndIsDirectory) {
            return AdapterError(
                code = if(existsAndIsDirectory != null) "EISDIR" else "ENOENT",
                path = path,
                syscall = "open"
            )
        }

        val file = File(itemPath)
        return if(utf8)
            file.readText(Charsets.UTF_8)
        else
            file.readBytes()
    }

    fun writeFile(path: String, data: ByteArray) : Any {
        val itemPath = this.baseDirectory + "/" + path

        val file = File(itemPath);

        try {
            file.writeBytes(data)
        } catch (e: Exception) {
            return AdapterError(
                code = "ENOENT",
                path = path,
                syscall = "open"
            )
        }

        return true
    }

    fun unlink(path: String): Boolean? {
        val itemPath = this.baseDirectory + "/" + path

        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == null || existsAndIsDirectory) return null

        val file = File(itemPath)
        file.delete()

        return true
    }

    fun readdir(path: String, withFileTypes: Boolean, recursive: Boolean): Any {
        val itemPath = this.baseDirectory + "/" + path

        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == null || !existsAndIsDirectory) {
            return AdapterError(
                code = if (existsAndIsDirectory != null) "ENOTDIR" else "ENOENT",
                path = path,
                syscall = "open"
            )
        }

        val dir = File(itemPath)
        val files = arrayListOf<File>()

        if(recursive)
            dir.walk().forEach { files.add(it) }
        else
            dir.listFiles()?.forEach { files.add(it) }

        if (!withFileTypes) return files.map { file -> file.name }

        val filesWithTypes = files.map { file -> mapOf(
            "name" to file.name,
            "isDirectory" to file.isDirectory
        )}

        return filesWithTypes
    }

    fun mkdir(path: String): Boolean {
        val itemPath = this.baseDirectory + "/" + path
        val dir = File(itemPath)
        dir.mkdirs()
        return true
    }

    fun rmdir(path: String): Boolean? {
        val itemPath = this.baseDirectory + "/" + path
        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == null || !existsAndIsDirectory) return null
        val dir = File(itemPath)
        dir.deleteRecursively()
        return true
    }

    fun stat(path: String) : Any {
        val itemPath = this.baseDirectory + "/" + path

        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)

        if(existsAndIsDirectory == null) {
            return AdapterError(
                code = "ENOENT",
                path = path,
                syscall = "stat"
            )
        }

        val item = Files.readAttributes(Paths.get(itemPath), BasicFileAttributes::class.java)

        return mapOf(
            "size" to item.size(),
            "isDirectory" to item.isDirectory,
            "isFile" to item.isRegularFile,
            "ctime" to item.creationTime().toString(),
            "ctimeMs" to item.creationTime().toMillis(),
            "mtime" to item.lastModifiedTime().toString(),
            "mtimeMs" to item.lastModifiedTime().toMillis()
        )
    }

    fun exists(path: String) : Any {
        val itemPath = this.baseDirectory + "/" + path

        val existsAndIsDirectory = this.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == null) return false

        return mapOf(
            "isFile" to !existsAndIsDirectory
        )
    }

}

File: platform/ios/xcode/FullStacked/Adapter.swift
//
//  Adapter.swift
//  FullStacked
//
//  Created by Charles-Philippe Lepage on 2024-03-22.
//

import Foundation
import UniformTypeIdentifiers
import SwiftyJSON

struct AdapterError {
    let code: String
    let path: String
    let syscall: String
    var toJSON: [String: Any] {
        return ["code": code, "path": path, "errno": -2, "syscall": syscall]
    }
}

class Adapter {
    let platform = "ios"
    let projectId: String?;
    var fs: AdapterFS
    
    init(projectId: String?, baseDirectory: String) {
        self.projectId = projectId
        self.fs = AdapterFS(baseDirectory: baseDirectory)
    }
    
    func callAdapterMethod(methodPath: [String.SubSequence], body: Data, done: @escaping  (_ maybeData: Any?) -> Void)  {
        if(methodPath.count == 0) {
            return done(nil)
        }
        
        let json = body.count == 0 ? JSON("[]") : try! JSON(data: body)
        
        let writeFile = { (path: String, data: Data, recursive: Bool) in
            if(recursive) {
                let directory = path.split(separator: "/").dropLast()
                self.fs.mkdir(path: directory.joined(separator: "/"))
            }
            
            return self.fs.writeFile(file: path, data: data)
        }
        
        switch(methodPath.first) {
        case "platform": return done(self.platform)
        case "fs":
            switch(methodPath[1]) {
            case "readFile": return done(self.fs.readFile(path: json[0].stringValue, utf8: json[1]["encoding"].stringValue == "utf8"))
            case "writeFile":
                var data: Data;
                
                if(json[1]["type"].stringValue == "Uint8Array") {
                    let uint8array = json[1]["data"].arrayValue.map({ number in
                        return number.uInt8!
                    })
                    data = Data(uint8array)
                } else {
                    data = json[1].stringValue.data(using: .utf8)!
                }
                
            return done(writeFile(json[0].stringValue, data, json[2]["recursive"].boolValue))
            case "writeFileMulti":
                for fileJSON in json[0].arrayValue {
                    var data: Data;
                    
                    if(fileJSON["data"]["type"].stringValue == "Uint8Array") {
                        let uint8array = fileJSON["data"]["data"].arrayValue.map({ number in
                            return number.uInt8!
                        })
                        data = Data(uint8array)
                    } else {
                        data = fileJSON["data"].stringValue.data(using: .utf8)!
                    }
                    
                    let maybeError = writeFile(fileJSON["path"].stringValue, data, json[1]["recursive"].boolValue)
                    if(maybeError is AdapterError){
                        return done(maybeError)
                    }
                }
                return done(true)
            case "unlink": return done(self.fs.unlink(path: json[0].stringValue))
            case "readdir": return done(self.fs.readdir(path: json[0].stringValue, withFileTypes: json[1]["withFileTypes"].boolValue, recursive: json[1]["recursive"].boolValue))
            case "mkdir": return done(self.fs.mkdir(path: json[0].stringValue))
            case "rmdir": return done(self.fs.rmdir(path: json[0].stringValue))
            case "stat": return done(self.fs.stat(path: json[0].stringValue))
            case "lstat": return done(self.fs.lstat(path: json[0].stringValue))
            case "exists":
                let exists = self.fs.exists(path: json[0].stringValue)
                return done(exists == nil ? false : exists)
            default: break
            }
            break
        case "fetch":
            var body: Data;
            
            if(json[1]["body"]["type"].stringValue == "Uint8Array") {
                let uint8array = json[1]["body"]["data"].arrayValue.map({ number in
                    return number.uInt8!
                })
                body = Data(uint8array)
            } else {
                body = json[1]["body"].stringValue.data(using: .utf8)!
            }
        
            let headersJSON = json[1]["headers"].dictionaryValue
            var headers: [String: String] = [:]
            headersJSON.keys.forEach { header in
                headers[header] = headersJSON[header]!.stringValue
            }
            
            return self.fetch(
                urlStr: json[0].stringValue,
                headers: headers,
                method: json[1]["method"].stringValue,
                timeout: json[1]["timeout"].double,
                body: body) { headers, statusCode, statusMessage, data in
                    var body: Any?
                    
                    if (json[1]["encoding"].stringValue == "utf8") {
                        body = String(data: data, encoding: .utf8)
                    } else {
                        body = ["type": "Uint8Array", "data": [UInt8](data)]
                    }
                    
                    DispatchQueue.main.async {
                        done([
                            "headers": headers,
                            "statusCode": statusCode,
                            "statusMessage": statusMessage,
                            "body": body
                        ])
                    }
                }
        case "broadcast":
            let peerMessage = [
                "projectId": self.projectId,
                "data": json[0].stringValue
            ]
            InstanceEditor.singleton!.push(messageType: "sendData", message: JSON(peerMessage).rawString()!)
            return done(true);
            
        default: break
        }
        
        return done(nil)
    }
    
    func fetch(urlStr: String,
               headers: [String: String],
               method: String,
               timeout: Double?,
               body: Data,
               onCompletion: @escaping (
                  _ headers: [String: String],
                  _ statusCode: Int,
                  _ statusMessage: String,
                  _ data: Data
               ) -> Void) {
                   let url = URL(string: urlStr)!
                   var request = URLRequest(url: url)
                   
                   request.httpMethod = method.isEmpty ? "GET" : method
                   
                   if(timeout != nil){
                       request.timeoutInterval = timeout! / 1000
                   }
                   
                   for (headerName, headerValue) in headers {
                       request.setValue(headerValue, forHTTPHeaderField: headerName)
                   }
                   
                   request.httpBody = body
                   
                   let task = URLSession.shared.dataTask(with: request) { data, response, error in
                       if error != nil {
                           onCompletion([:], 500, "Fetch error", Data())
                           return
                       }
                       
                       let headers = (response as! HTTPURLResponse).allHeaderFields as! [String: String]
                       let statusCode = (response as! HTTPURLResponse).statusCode
                       let statusMessage = "OK"
                       
                       onCompletion(headers, statusCode, statusMessage, data ?? Data())
                   }
                   task.resume()
               }
}

class AdapterFS {
    // @returns {nil} if doesn't exists, {true} if exists and directory, {false} if exists and file
    static func itemExistsAndIsDirectory (_ path: String) -> Bool? {
        var isDirectory: ObjCBool = false
        let exists = FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory)
        return exists ? isDirectory.boolValue : nil
    }
    
    static func mimeType(filePath: String) -> String {
        if let mimeType = UTType(filenameExtension: (filePath as NSString).pathExtension)?.preferredMIMEType {
            return mimeType
        }
        else {
            return "application/octet-stream"
        }
    }
    
    let baseDirectory: String
    
    init(baseDirectory: String){
        self.baseDirectory = baseDirectory
    }
    
    func readFile(path: String, utf8: Bool) -> Any {
        let itemPath = self.baseDirectory + "/" + path
        
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath);
        if(existsAndIsDirectory == nil || existsAndIsDirectory!) {
            return AdapterError(
                code: existsAndIsDirectory != nil ? "EISDIR" : "ENOENT",
                path: path,
                syscall: "open"
            )
        }
        
        let contents = FileManager.default.contents(atPath: itemPath)!
        
        if(utf8){
            return String(data: contents, encoding: .utf8)!
        }
        
        return contents
    }
    
    func writeFile(file: String, data: Data) -> Any? {
        let itemPath = self.baseDirectory + "/" + file
        
        do {
            try data.write(to: URL(fileURLWithPath: itemPath))
        } catch {
            return AdapterError(
                code: "ENOENT",
                path: file,
                syscall: "open"
            )
        }
        
        return true
    }
    
    func unlink(path: String) {
        let itemPath = self.baseDirectory + "/" + path
        
        // let's at least try to act like nodejs unlink and not delete directories
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath)
        let isFile = existsAndIsDirectory != nil && !existsAndIsDirectory!
        if(isFile) {
            try! FileManager.default.removeItem(atPath: itemPath)
        }
    }
    
    func readdir(path: String, withFileTypes: Bool, recursive: Bool) -> Any {
        let itemPath = self.baseDirectory + "/" + path;
        
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath);
        if(existsAndIsDirectory == nil || !existsAndIsDirectory!) {
            return AdapterError(
                code: existsAndIsDirectory != nil ? "ENOTDIR" : "ENOENT",
                path: path,
                syscall: "open"
            )
        }
        
        var items = recursive
            ? []
            : try! FileManager.default.contentsOfDirectory(atPath: itemPath)
        
        if(recursive) {
            let enumarator = FileManager.default.enumerator(atPath: itemPath)
            while let element = enumarator?.nextObject() as? String {
                items.append(element)
            }
        }
        
        if(withFileTypes){
            let itemsWithFileTypes = items.map { childItem in
                var isDirectory: ObjCBool = false
                let childItemPath = itemPath + "/" + childItem
                FileManager.default.fileExists(atPath: childItemPath, isDirectory: &isDirectory)
                return ["name": childItem, "isDirectory": isDirectory.boolValue]
            }
            
            return itemsWithFileTypes
        }
        
        return items
    }
    
    func mkdir(path: String) {
        let itemPath = self.baseDirectory + "/" + path
        try! FileManager.default.createDirectory(atPath: itemPath, withIntermediateDirectories: true)
    }
    
    func rmdir(path: String) {
        let itemPath = self.baseDirectory + "/" + path
        
        // let's at least try to act like nodejs rmdir and delete only directories
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath);
        if(existsAndIsDirectory != nil && existsAndIsDirectory!) {
            try! FileManager.default.removeItem(atPath: itemPath)
        }
    }
    
    func stat (path: String) -> Any {
        let itemPath = self.baseDirectory + "/" + path;
        
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath);
        if(existsAndIsDirectory == nil) {
            return AdapterError(
                code: "ENOENT",
                path: path,
                syscall: "stat"
            )
        }
        
        let stats = try! FileManager.default.attributesOfItem(atPath: itemPath)
        
        return [
            "size": stats[FileAttributeKey.size],
            "isDirectory": existsAndIsDirectory!,
            "isFile": !existsAndIsDirectory!,
            "ctime": (stats[FileAttributeKey.creationDate] as! Date).ISO8601Format() ,
            "ctimeMs": (stats[FileAttributeKey.creationDate] as! Date).timeIntervalSince1970 * 1000,
            "mtime": (stats[FileAttributeKey.modificationDate] as! Date).ISO8601Format(),
            "mtimeMs": (stats[FileAttributeKey.modificationDate] as! Date).timeIntervalSince1970 * 1000
        ]
    }
    
    func lstat(path: String) -> Any {
        return self.stat(path: path)
    }
    
    func exists(path: String) -> Any? {
        let itemPath = self.baseDirectory + "/" + path
        
        let existsAndIsDirectory = AdapterFS.itemExistsAndIsDirectory(itemPath)
        if(existsAndIsDirectory == nil){
            return nil
        }
        
        return [
            "isFile": !existsAndIsDirectory!
        ]
    }
}
